File: apps/patter-voice-sdk/libraries/python/getpatter/providers/telnyx_adapter.py
"""Telnyx :class:`TelephonyProvider` — number provisioning and call control.

Thin async wrapper over Telnyx's v2 REST API used by :mod:`getpatter.client`
to provision numbers, place outbound calls and hang up. Media streaming is
attached separately from ``call.answered`` (see :mod:`telephony.telnyx`).
"""

import logging

import httpx

from getpatter.providers.base import TelephonyProvider

logger = logging.getLogger("getpatter.providers.telnyx_adapter")

TELNYX_API_BASE = "https://api.telnyx.com/v2"


class TelnyxAdapter(TelephonyProvider):
    """:class:`TelephonyProvider` implementation backed by Telnyx Call Control."""

    def __init__(self, api_key: str, connection_id: str = ""):
        self.api_key = api_key
        self.connection_id = connection_id
        self._client = httpx.AsyncClient(
            base_url=TELNYX_API_BASE,
            headers={"Authorization": f"Bearer {api_key}"},
            timeout=30.0,
        )

    def __repr__(self) -> str:
        return f"TelnyxAdapter(connection_id={self.connection_id!r})"

    async def __aenter__(self) -> "TelnyxAdapter":
        return self

    async def __aexit__(self, *args: object) -> None:
        await self.close()

    async def provision_number(self, country: str) -> str:
        """Search and order an available Telnyx number for the given ISO country."""
        # Telnyx search filter uses nested ``filter[phone_number][country_code]``
        # (not ``filter[country_code]``). See:
        # https://developers.telnyx.com/api/numbers/list-available-phone-numbers
        resp = await self._client.get(
            "/available_phone_numbers",
            params={
                "filter[phone_number][country_code]": country,
                "filter[limit]": 1,
            },
        )
        resp.raise_for_status()
        numbers = resp.json()["data"]
        if not numbers:
            raise ValueError(f"No numbers for {country}")
        chosen = numbers[0]["phone_number"]
        order_body: dict = {"phone_numbers": [{"phone_number": chosen}]}
        # When a Call Control Application is bound to this adapter, attach
        # it to the order so the newly-purchased number is ready to place
        # and receive calls without a follow-up PATCH.
        if self.connection_id:
            order_body["connection_id"] = self.connection_id
        buy = await self._client.post("/number_orders", json=order_body)
        buy.raise_for_status()
        return chosen

    async def configure_number(self, number: str, webhook_url: str) -> None:
        """Associate number with the Call Control Application.

        ``connection_id`` lives on ``PATCH /phone_numbers/{id}`` — the
        ``/voice`` sub-resource only covers voice settings and silently
        ignores unknown body fields, so the previous single ``/voice`` PATCH
        returned 200 without ever linking the number to the Call Control app
        (inbound calls didn't route until fixed in the portal). Send the
        association to the base endpoint and keep voice settings on
        ``/voice``.

        ``number`` may be the phone_number ID or the E.164 string; the API
        accepts both identifiers but the phone_number ID is preferred.
        """
        from urllib.parse import quote as _quote

        encoded = _quote(number, safe="")
        assoc = await self._client.patch(
            f"/phone_numbers/{encoded}",
            json={"connection_id": self.connection_id},
        )
        if assoc.status_code >= 400:
            import logging as _logging

            _logging.getLogger("getpatter").warning(
                "Telnyx connection association returned %s: %s",
                assoc.status_code,
                assoc.text[:300],
            )
        payload = {
            "tech_prefix_enabled": False,
        }
        resp = await self._client.patch(
            f"/phone_numbers/{encoded}/voice",
            json=payload,
        )
        if resp.status_code >= 400:
            # Surface the server-side error body so misconfigured
            # connection_ids / unknown numbers don't fail silently.
            import logging as _logging

            _logging.getLogger("getpatter").warning(
                "Telnyx configure_number returned %s: %s",
                resp.status_code,
                resp.text[:300],
            )
        resp.raise_for_status()

    async def initiate_call(
        self,
        from_number: str,
        to_number: str,
        stream_url: str,
        *,
        ring_timeout: int | None = None,
        client_state: str | None = None,
        machine_detection: bool = False,
    ) -> str:
        """Place an outbound Call Control dial.

        NOTE: ``stream_url`` / ``stream_track`` are NOT accepted by
        ``POST /calls``. Telnyx attaches media streaming after the call is
        answered via ``POST /calls/{id}/actions/streaming_start``, triggered
        from the ``call.answered`` webhook for outgoing legs (the inbound
        path folds streaming into ``actions/answer`` on ``call.initiated``).
        The ``stream_url`` argument is retained for interface parity
        (``TelephonyProvider``) but is intentionally unused here.

        See the ``call.answered`` handler in ``server.py`` for the full flow.

        Args:
            from_number: Caller E.164 number (must be owned on the connection).
            to_number: Callee E.164 number.
            stream_url: Unused — media stream is attached after ``call.answered``.
            ring_timeout: Max seconds to wait before the leg is marked no-answer.
            client_state: Optional opaque string echoed back on every webhook
                for this call. Encoded as base64 per Telnyx contract.
        """
        del stream_url  # see docstring — retained only for TelephonyProvider parity
        payload: dict = {
            "connection_id": self.connection_id,
            "from": from_number,
            "to": to_number,
        }
        if ring_timeout is not None:
            payload["timeout_secs"] = int(ring_timeout)
        if machine_detection:
            # ``greeting_end`` is the production-recommended mode: Telnyx
            # returns the human/machine classification on
            # ``call.machine.detection.ended`` AND emits a follow-up
            # ``call.machine.greeting.ended`` once the answering-machine
            # greeting reaches the beep, so a downstream voicemail-drop
            # can speak immediately after the prompt.
            payload["answering_machine_detection"] = "greeting_end"
        if client_state:
            # Telnyx expects client_state as base64-encoded opaque string.
            import base64 as _b64

            payload["client_state"] = _b64.b64encode(
                client_state.encode("utf-8")
            ).decode("ascii")
        resp = await self._client.post("/calls", json=payload)
        resp.raise_for_status()
        return resp.json()["data"]["call_control_id"]

    async def end_call(self, call_id: str, *, command_id: str | None = None) -> None:
        """Hang up an active Telnyx call.

        ``command_id`` provides idempotency on retries (Telnyx will ignore a
        duplicate action with the same ``command_id``). When omitted, a
        UUID4 is generated automatically.
        """
        import uuid as _uuid
        from urllib.parse import quote as _quote

        body = {"command_id": command_id or str(_uuid.uuid4())}
        await self._client.post(
            f"/calls/{_quote(call_id, safe='')}/actions/hangup", json=body
        )

    def record_call_end_cost(self, *, duration_seconds: float, direction: str) -> None:
        """Emit ``patter.cost.telephony_minutes`` on the active span.

        Called by the embedded server's bridge cleanup once the call's
        wall-clock duration is known.
        """
        try:
            from getpatter.observability.attributes import record_patter_attrs

            record_patter_attrs(
                {
                    "patter.cost.telephony_minutes": duration_seconds / 60.0,
                    "patter.telephony": "telnyx",
                    "patter.direction": direction,
                }
            )
        except Exception:  # pragma: no cover — defense in depth
            logger.debug("record_call_end_cost failed", exc_info=True)

    async def close(self) -> None:
        """Close the underlying HTTP client."""
        await self._client.aclose()

File: apps/patter-voice-sdk/libraries/typescript/src/providers/telnyx-adapter.ts
/**
 * Telnyx telephony adapter — parity with Python ``TelnyxAdapter``.
 *
 * This adapter talks directly to Telnyx's REST API via ``fetch`` (Bearer auth)
 * and avoids the vendor SDK dependency.
 *
 * Wave 1 fixes applied (aligned with Python Wave1-Team7):
 *   - Do NOT pass ``stream_url`` when dialing — streams are attached via the
 *     Call Control Application, not per call.
 *   - Use ``/number_orders`` with a ``filter[phone_number]`` param on the
 *     ``/available_phone_numbers`` lookup so we reserve specifically the
 *     number we intend to buy.
 *   - Fully URL-encode ``callControlId`` in every path interpolation.
 *
 * See: ``libraries/python/getpatter/providers/telnyx_adapter.py``.
 */
import { randomUUID } from 'node:crypto';
import { getLogger } from '../logger';

const TELNYX_API_BASE = 'https://api.telnyx.com/v2';

/** Options accepted by {@link TelnyxAdapter.provisionNumber}. */
export interface ProvisionNumberOptions {
  /** ISO-3166-1 alpha-2 country code (e.g. ``"US"``). */
  countryCode: string;
}

/** Result returned by {@link TelnyxAdapter.provisionNumber}. */
export interface ProvisionNumberResult {
  readonly phoneNumber: string;
  readonly orderId: string;
}

/** Options accepted by {@link TelnyxAdapter.configureNumber}. */
export interface ConfigureNumberOptions {
  /** Telnyx Call Control Application / Connection ID. */
  connectionId: string;
}

/** Options accepted by {@link TelnyxAdapter.initiateCall}. */
export interface InitiateCallOptions {
  from: string;
  to: string;
  /** Override ``connectionId`` at dial time. Falls back to the adapter default. */
  connectionId?: string;
  /** Opaque state string that Telnyx echoes back on webhooks. Base64-encoded on wire. */
  clientState?: string;
}

/** Result returned by {@link TelnyxAdapter.initiateCall}. */
export interface InitiateCallResult {
  readonly callControlId: string;
}

/** Options accepted by {@link TelnyxAdapter.endCall}. */
export interface EndCallOptions {
  /** Idempotency key for the hangup command. */
  commandId?: string;
}

interface TelnyxAvailableNumbersPayload {
  data?: Array<{ phone_number?: string }>;
}

interface TelnyxNumberOrderPayload {
  data?: { id?: string };
}

interface TelnyxCallPayload {
  data?: { call_control_id?: string };
}

/** Direct REST adapter for Telnyx Call Control & Numbers API. */
export class TelnyxAdapter {
  private readonly apiKey: string;
  readonly connectionId: string | undefined;
  private readonly baseUrl: string = TELNYX_API_BASE;

  constructor(apiKey: string, connectionId?: string) {
    if (!apiKey) throw new Error('TelnyxAdapter: apiKey is required');
    this.apiKey = apiKey;
    this.connectionId = connectionId;
  }

  private async request<T>(
    method: 'GET' | 'POST' | 'PATCH',
    path: string,
    body?: Record<string, unknown>,
  ): Promise<T> {
    const url = `${this.baseUrl}${path}`;
    const headers: Record<string, string> = {
      Authorization: `Bearer ${this.apiKey}`,
    };
    if (body !== undefined) headers['Content-Type'] = 'application/json';

    const response = await fetch(url, {
      method,
      headers,
      body: body !== undefined ? JSON.stringify(body) : undefined,
      signal: AbortSignal.timeout(30_000),
    });

    const text = await response.text();
    if (!response.ok) {
      throw new Error(`Telnyx ${method} ${path} failed: ${response.status} ${text}`);
    }
    if (!text) return {} as T;
    try {
      return JSON.parse(text) as T;
    } catch (e) {
      throw new Error(`Telnyx returned non-JSON response: ${String(e)}`);
    }
  }

  /**
   * Search available numbers for ``countryCode`` and place an order for the
   * first match. Returns both the reserved E.164 number and the order ID.
   */
  async provisionNumber(opts: ProvisionNumberOptions): Promise<ProvisionNumberResult> {
    const country = encodeURIComponent(opts.countryCode);
    // Telnyx search filter uses nested ``filter[phone_number][country_code]``
    // (not ``filter[country_code]``). See the Python adapter and the
    // telnyx-numbers-compliance skill.
    const searchPath =
      `/available_phone_numbers?filter[phone_number][country_code]=${country}&filter[limit]=1`;
    const available = await this.request<TelnyxAvailableNumbersPayload>('GET', searchPath);
    const chosen = available.data?.[0]?.phone_number;
    if (!chosen) {
      throw new Error(`TelnyxAdapter: no numbers available for ${opts.countryCode}`);
    }
    // When a Call Control Application is bound to this adapter, attach it to
    // the order so the newly-purchased number is immediately ready to place
    // and receive calls without a follow-up PATCH.
    const orderBody: Record<string, unknown> = {
      phone_numbers: [{ phone_number: chosen }],
    };
    if (this.connectionId) {
      orderBody.connection_id = this.connectionId;
    }
    const order = await this.request<TelnyxNumberOrderPayload>(
      'POST',
      '/number_orders',
      orderBody,
    );
    const orderId = order.data?.id;
    if (!orderId) throw new Error('TelnyxAdapter: /number_orders returned no order id');
    return { phoneNumber: chosen, orderId };
  }

  /** Attach a number to a Call Control Application. */
  async configureNumber(
    phoneNumber: string,
    opts: ConfigureNumberOptions,
  ): Promise<void> {
    if (!phoneNumber) throw new Error('TelnyxAdapter: phoneNumber is required');
    if (!opts.connectionId) throw new Error('TelnyxAdapter: connectionId is required');
    // ``connection_id`` lives on ``PATCH /phone_numbers/{id}`` — the
    // ``/voice`` sub-resource only covers voice settings and silently
    // ignores unknown fields, so the previous single ``/voice`` PATCH
    // returned 200 without ever linking the number to the Call Control app.
    // ``phoneNumber`` may be the phone_number ID or the E.164 string.
    try {
      await this.request<unknown>(
        'PATCH',
        `/phone_numbers/${encodeURIComponent(phoneNumber)}`,
        { connection_id: opts.connectionId },
      );
      await this.request<unknown>(
        'PATCH',
        `/phone_numbers/${encodeURIComponent(phoneNumber)}/voice`,
        { tech_prefix_enabled: false },
      );
    } catch (err) {
      // Re-throw with a sanitised message that omits the phone number from the
      // path to prevent E.164 PII leaking through error logs.
      const status = err instanceof Error ? err.message.replace(/\+\d{7,15}/g, '[REDACTED]') : String(err);
      throw new Error(`TelnyxAdapter: configureNumber failed: ${status}`);
    }
  }

  /**
   * Place an outbound call on the Call Control Application.
   *
   * Note: we intentionally do NOT pass ``stream_url`` here — audio streaming
   * is configured on the Application itself (or started explicitly via a
   * ``streaming_start`` command). Passing ``stream_url`` on dial is a
   * deprecated code path that Telnyx rejects in newer API versions.
   */
  async initiateCall(opts: InitiateCallOptions): Promise<InitiateCallResult> {
    const connectionId = opts.connectionId ?? this.connectionId;
    if (!connectionId) {
      throw new Error('TelnyxAdapter: connectionId must be provided to initiateCall');
    }
    const payload: Record<string, unknown> = {
      connection_id: connectionId,
      from: opts.from,
      to: opts.to,
    };
    if (opts.clientState) {
      payload.client_state = Buffer.from(opts.clientState, 'utf-8').toString('base64');
    }
    const resp = await this.request<TelnyxCallPayload>('POST', '/calls', payload);
    const callControlId = resp.data?.call_control_id;
    if (!callControlId) {
      throw new Error('TelnyxAdapter: /calls returned no call_control_id');
    }
    return { callControlId };
  }

  /** Hang up an in-progress call. */
  async endCall(callControlId: string, opts: EndCallOptions = {}): Promise<void> {
    if (!callControlId) throw new Error('TelnyxAdapter: callControlId is required');
    const encoded = encodeURIComponent(callControlId);
    // ``command_id`` provides idempotency on retries (Telnyx will ignore a
    // duplicate action with the same ``command_id``). When omitted we
    // auto-generate a UUID4 to match the Python adapter's behaviour.
    const body: Record<string, unknown> = {
      command_id: opts.commandId ?? randomUUID(),
    };
    try {
      await this.request<unknown>(
        'POST',
        `/calls/${encoded}/actions/hangup`,
        body,
      );
    } catch (err) {
      getLogger().warn(
        `[TelnyxAdapter] endCall failed for ${callControlId}: ${String(err)}`,
      );
      throw err;
    }
  }
}
